core/types/calls.cc

```diff
@@ -2395,20 +2396,52 @@ class Array_flatten : public IntrinsicMethod {
             // This only shows up because t->elementType(gs) for tuples returns an OrType of all its elements.
             // So to properly handle nested tuples, we have to descend into the OrType's.
             [&](const OrType &o) {
-                result = Types::any(gs, recursivelyFlattenArrays(gs, o.left, newDepth),
-                                    recursivelyFlattenArrays(gs, o.right, newDepth));
+                result = Types::any(gs, recursivelyFlattenArrays(gs, args, o.left, newDepth),
+                                    recursivelyFlattenArrays(gs, args, o.right, newDepth));
+            },
+
+            [&](const ClassType &c) {
+                // If the element type supports the #to_ary method, then Ruby will implicitly call it when
+                // flattening. So here we need to get the return type of calling #to_ary on an element.
+                DispatchArgs innerArgs{Names::toAry(), args.locs, 0,
+                                       args.args,      type,      {type, args.fullType.origins},
+                                       type,           nullptr,   args.originForUninitialized};
```

Comment:
This is the right idea, but the call to `to_ary` that you're dispatching doesn't have the arguments of the original call to `flatten`, it has zero arguments.  So you need to set up things for a zero-argument call.  `Magic_splat.apply` shows how this is done.